File: Problems/0374-guess-number-higher-or-lower-ii.py
# LeetCode solution starts from here

class Solution(object):
    def getMoneyAmount(self, n):
        dp = [[0] * (n + 2) for _ in range(n + 2)]

        for length in range(2, n + 1):
            for left in range(1, n - length + 2):
                right = left + length - 1
                dp[left][right] = float('inf')

                for x in range(left, right + 1):
                    dp[left][right] = min(
                        dp[left][right],
                        x + max(dp[left][x - 1], dp[x + 1][right])
                    )

        return dp[1][n]

File: Problems/0374-guess-number-higher-or-lower-ii.cpp
#include <bits/stdc++.h>
using namespace std;

// LeetCode solution starts from here
class Solution {
public:
    int guessNumber(int n) {
        long long left = 1, right = n;

        while (left <= right) {
            long long mid = left + (right - left) / 2;
            int res = guess(mid);

            if (res == 0)
                return mid;
            else if (res < 0)
                right = mid - 1;
            else
                left = mid + 1;
        }

        return -1;
    }
};